### PR TITLE
Dnsmadeeasy provider

### PR DESCRIPTION
**Pull request by yzguy**
Figure I'd get this up here to get some feedback while I work on the tests (~73% coverage right now)

Add DNSMadeEasy provider

* No support for Geo or ANAME records
* Allows user to toggle sandbox environment
* Added zone creation detection (#175)

I have notes for investigating paging, I added 2500 records and I could not get a response that had `totalPages > 1`. It was always 1 response, with all the records. I'm not sure what the limit is where it will start paging. Opening a support ticket with DnsMadeEasy to find out if there is pagination at all.

Fixes: #150

### CODE REVIEW

**Comment by yzguy** [comment]
Response from their support:

```
This feature was added based on our users request. By default we will return all records, but you do have the ability to paginate the responses as well in smaller chunks if necessary.

Example:
https://api.dnsmadeeasy.com/V2.0/dns/managed/<domain_id>/records/?rows=5&page=4
https://api.dnsmadeeasy.com/V2.0/dns/managed/<domain_id>/records/?rows=5&page=3
```

So is it worth implementing pagination now to avoid an issue due to large response size, etc down the road.

File: octodns/provider/dnsmadeeasy.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    PRODUCTION = 'https://api.dnsmadeeasy.com/V2.0/dns/managed'
+    SANDBOX = 'https://api.sandbox.dnsmadeeasy.com/V2.0/dns/managed'
+
+    def __init__(self, api_key, secret_key, sandbox=False,
+                 ratelimit_delay=0.0):
+        self.api_key = api_key
+        self.secret_key = secret_key
+        self._base = self.PRODUCTION
+        if sandbox:
+            self._base = self.SANDBOX
```
Comment (ross) [inline octodns/provider/dnsmadeeasy.py]: These 3 lines can optionally be combined into `self._base = self.SANDBOX if sandbox else self.PRODUCTION`

File: octodns/provider/dnsmadeeasy.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    def _request(self, method, path, params=None, data=None):
+        now = self._current_time()
+        hmac_hash = self._hmac_hash(now)
+
+        headers = {
+            'x-dnsme-apiKey': self.api_key,
+            'x-dnsme-hmac': hmac_hash,
+            'x-dnsme-requestDate': now
+        }
+        self._sess.headers.update(headers)
```
Comment (ross) [inline octodns/provider/dnsmadeeasy.py]: Some of these look like they're session-level and others request specific. 

You can probably do `self._sess.headers.update({'x-dnsme-apiKey': api_key})` up in the `__init__` and avoid having to keep it around as an object property. 

The the `hmac` and `requestDate` can then be passed as `headers=headers` down in the `self._sess.request` call.

**Comment by yzguy** [comment]
Make changes :) Thanks for the feedback. Just working on tests now.

What are your thoughts on implementing pagination mentioned above, rather than just a "give me all the records". Do you think there could be any issue in the future with large zones + huge response from API?

**Comment by ross** [comment]
> What are your thoughts on implementing pagination mentioned above, rather than just a "give me all the records". Do you think there could be any issue in the future with large zones + huge response from API?

So there's pagination support, but the default page size is really large or something and you haven't hit it? Don't have a strong opinion in general. People are pretty good at hitting edge cases and have definitely found pagination problems in other providers so there's at least some chance that'll happen here, but if their API doesn't give any indication as to what it's size/behaviors are it might be fine to wait and see if/when someone hits it.

**Comment by yzguy** [comment]
That's kinda what I'm thinking, it should be easy enough to add it in around where its grabbing the records on the first response if it becomes an issue. It's nice having 1 request for all of them, because there is rate limiting on their API.

**Comment by yzguy** [comment]
After months of procrastinating, I have finally finished this first iteration of the DnsMadeEasy provider.

Tests are finished, all passing, and 100% test coverage.

Limitations currently are:
* No ANAME support
* No Geo Support

**Comment by ross** [comment]
> After months of procrastinating, I have finally finished this first iteration of the DnsMadeEasy provider.

😁 Cool. Will take a look at things and review it this evening.

**Comment by ross** [comment]
This was the last thing I was holding off for to land before cutting a new release. Will look to do that this week at some point.

**Comment by yzguy** [comment]
Thank you for your help :) Really glad I'm able to contribute, definitely a project I've very interested in.

**Pull request merged**